### ISSUE TITLE
How to obtain all minimum cuts via CLI

### ISSUE THREAD

**Issue by bzfhille** [VieCut/VieCut#6]
First of all, thank you very much for providing an implementation of an algorithm to enumerate all minimum cuts. This is exactly what I need in my application.

Unfortunatley, I was not able to obtain all minimum cuts via the CLI. According to the documentation, this should be possible using `mincut cactus`. However, this provides only one cut. I also didn't find a way to obtain the cactus representation to extract the cuts myself. For completeness, I am using the command

	mincut <metis_file> cactus -s

I am actually interested in some kind of balanced cut, however the balancing condition is according to some weights on the edges. It would be most convenient to have a list of all minimum cuts and select the best one in my Python code. Is there a way to obtain this without hacking the source code to provide that output?

**Comment by alexnoe** [comment]
Sorry, there currently exists no function to write the minimum cut cactus to file. 

Note that there are two separate problems of 'finding all minimum cuts':

(1) Finding the minimum s-t-cut between each pair of vertices s and t.
(2) Find all global minimal cuts (i.e. if the smallest cut has a weight of X, find all cuts of weight X)

This solves (2) and returns a cactus graph that describes the cuts and partitions defined by these cuts. There currently exists no way to write this to file. I'd be happy to accept a pull request if you were to add one!

**Comment by schulzchristian** [comment]
@bzfhille which problem do you need? The ``real'' mincut problem or the s-t min cut problem?

**Comment by bzfhille** [comment]
Good to know that I did not miss anything.

I'm actually interested in variant (2). I am happy to hack up a solution to my issue, likely by just writing all cuts to a file, based on the routine to determine the most balanced cut (I had a quick look into this already). I think exporting the cactus graph and implementing the logic to interpret that on my side is much more work.

**Comment by schulzchristian** [comment]
@alexnoe maybe you can give some concrete pointers. 
Outputting each cut in a separate file would be reasonable I guess.

**Comment by alexnoe** [comment]
The thing to be careful about when writing the cactus to file is that the cactus graph might represent more than n minimum cuts. For example, in a ring graph there are O(n^2) minimum cuts (each set of two edges in the ring), so if you want to enumerate all cuts to find the best one, staying within the DFS framework that looks at all cuts might be useful.

**Comment by bzfhille** [comment]
Thanks for providing the pointers. In the meantime, I had a look into the code and also in some papers. I tried to find out how to extract the minimum cuts from the cactus graph - and hopefully to find the corresponding code. However, if I understood your paper correctly, you implemented only the special case for determining the most balanced cut, which can be done more efficiently. Correct?

The papers also seem to focus on constructing the cactus graph; at least I did find an explicit description of the algorithm for extracting the minimum cuts. Of course, the definition of the cactus graph representation implies that all minimum cuts of the original graph can be obtained by enumerating all minimum cuts of the cactus. Given the special structure of a cactus, this amounts to considering each tree edge and each pair of edges of each cycle. Is this the way this should be done (via the "DFS framework" you mentioned)? If so, is there any code in VieCut that can be useful to implement this?

**Comment by schulzchristian** [comment]
@alexnoe can you comment on this?

**Comment by alexnoe** [comment]
Sorry, I had forgotten to respond, work has been stressful.

https://github.com/VieCut/VieCut/blob/master/lib/algorithms/global_mincut/cactus/balanced_cut_dfs.h#L70

Above is the starting point for the recursive DFS algorithm that we use to enumerate all cuts. We explore the graph in DFS order and whenever we encounter a new vertex t, we check whether the cut between t and its parent (in the DFS tree) is 'better' than the best cut we found so far (for us, that is more balanced, https://github.com/VieCut/VieCut/blob/master/lib/algorithms/global_mincut/cactus/balanced_cut_dfs.h#L87). This will reach all tree edges. If we encounter a vertex that we have seen already, we know that we are in a cycle, which we mark and then investigate when our DFS recursion leaves the cycle again (https://github.com/VieCut/VieCut/blob/master/lib/algorithms/global_mincut/cactus/balanced_cut_dfs.h#L144) - as we only aim to find the most balanced cut, we can do that in O(numVerticesInCycle), but depending on your function you might have to check all pairs of cycle edges.

So, in short:
tree edge examination code should be put here: https://github.com/VieCut/VieCut/blob/master/lib/algorithms/global_mincut/cactus/balanced_cut_dfs.h#L87
cycle examination code should be put here: https://github.com/VieCut/VieCut/blob/master/lib/algorithms/global_mincut/cactus/balanced_cut_dfs.h#L157 (cycle_vertices contains the list of vertices [of the cactus graph] that are in the cycle)

Otherwise you should be able to keep the existing DFS algorithm.

**Comment by bzfhille** [comment]
Thanks, this comment helped a lot!

By now, I added code to write the cactus to a file in GraphML format, which I can conveniently handle in python using NetworkX. Based on your explanations above, I was able to extract the cycles of the cactus graph via DFS. However, I was not able to determine the tree edges within the DFS procedure (I do have a workaround, so no big issue here). The reason is that the if condition in https://github.com/VieCut/VieCut/blob/master/lib/algorithms/global_mincut/cactus/balanced_cut_dfs.h#L87
is also true when visiting new nodes along a cycle. I probably need to have a closer look here, I had only a few hours to spent on this today.

**Comment by bzfhille** [comment]
In one of my simple tests cases (a K_4), the `containedVertices` of all nodes of the computed cactus are empty. Can this be OK? I expected the union of the contained vertices to be the node set of the original graph...

**Comment by alexnoe** [comment]
To comment 1: Yeah, that is a good observation. If I got this correctly, this should be implicitly the cut that separates all the already-explored cycle vertices (and their parents / ancestors), from the unexplored part of the graph. We will look at this cut (and all other cuts of the cycle later in line 144ff, but correct, we already look at that cut earlier)

To comment 2: that should not be the case, the cactus graph for a K_4 should be a star (with edge weight 3) with 4 'leafs', where each leaf contains one of the original vertex and the center vertex does not contain any vertices. Can you post your graph input file, maybe there is an issue there?

**Comment by schulzchristian** [comment]
any update on this?

**Comment by alexnoe** [comment]
Hi Christian,

we managed to solve the issue above in a separate ticket: https://github.com/VieCut/VieCut/issues/7 that has since been closed. I do not know of any further issue. I am leaving this ticket open as a general issue-tracker.

**Comment by schulzchristian** [comment]
It would be good to integrate this functionality in the main branch as there are a couple of people who seem to need it.

**Comment by alexnoe** [comment]
I'd be happy to merge any CR.

**Comment by bzfhille** [comment]
Unfortunately, I didn't have much time to work on this side project recently. The status is that I added code for writing the cactus graph to an GraphML file - see the new PR https://github.com/VieCut/VieCut/pull/8 Please provide feedback on what else is needed to get this accepted.

I'm still working on external Python code to use this cactus graph to enumerate the cuts I am interested in.

**Comment by alexnoe** [comment]
Looks good to me, squash-merged it into mainline. Thank you very much!

**Comment by schulzchristian** [comment]
looks like this is finished. I am closing the issue now.
